Reorganization to reduce duplicate code

warnerjon12 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Partially fixes #10047

#### Brief description of what is fixed or changed
Reorganized some of the code to reduce duplicate code and encourage code reuse

#### Other comments
Makes the code more maintainable

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

warnerjon12 committed:
Reimplemented expression comparison to remove duplicate code

Undefined infinite comparisons (e.g. oo > oo) now raise ValueError.
Formerly, this would have returned a true or false value.

warnerjon12 committed:
Reimplemented _eval_is_extended_positive/negative to reduce duplicate code

The methods were identical except one expression.
Each now calls a wrapper that takes a 'positive' parameter.

warnerjon12 committed:
Reimplemented part of _eval_interval to reduce duplicate code

warnerjon12 committed:
Reimplemented part of Expr.coeff to reduce duplicate code

warnerjon12 committed:
Fixed a small issue in Expr.extract_multiplicatively

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -329,98 +329,55 @@ def __complex__(self):
         n2 = _n2(self, other)
         if n2 is not None:
-            return _sympify(n2 >= 0)
+            return eval("_sympify(n2 %s 0)" % op)
```
We shouldn't use eval at all here.

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -329,98 +329,55 @@ def __complex__(self):
+                test = eval("diff.%s" %
+                            {"<": "is_extended_negative",
+                             ">": "is_extended_positive",
+                             "<=": "is_extended_nonpositive",
```
We shouldn't use eval here either. There might be a good way of doing this with lambda functions or with operator.attrgetter and operator.le etc.

asmeurer replied:
Just do

```py
if op == '<':
    test = diff.is_extended_negative
elif op == '>':
    test = diff.is_extended_positive
...
```

You lose the value of removing code duplication if you make the code harder to read and more bug prone.

warnerjon12 replied:
Okay, I changed it

asmeurer commented:
Let's make sure the coverage on this code is at 100% so that we are sure no new bugs are introduced.

oscarbenjamin commented:
The code you are editing here is an absolutely central part of sympy that is used heavily. We will need to check carefully that for slowdowns as well as correctness.

It looks as if some of the changes might be fixing bugs. If so tests should be added for cases that fail on current master.

warnerjon12 committed:
Fixed test for issue #10047, which is now fixed

warnerjon12 committed:
Updated test for issue #10047

warnerjon12 committed:
Changes for improved readability in Expr._cmp

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -329,98 +329,66 @@ def __complex__(self):
+                if test is not None:
+                    return sympify(test)
+            else: # e.g. oo > oo
+                raise ValueError("Undefined comparison of infinite values")
```
What happens for `oo <= oo`?

warnerjon12 replied:
`oo <= oo` is equivalent to `oo - oo <= 0` which should result in the ValueError being raised

warnerjon12 replied:
I will see about adding a few tests for code coverage

warnerjon12 replied:
I suppose treating `oo <= oo` as an undefined expression is different than the current implementation, but I feel like it is the correct thing to do. Mathematically, `oo` represents the limit of some real-valued thing, so the value of comparing two infinities of the same sign is undefined unless we know something about how fast each thing grows in the limit

oscarbenjamin replied:
We are using the extended real numbers so we should have `oo <= oo`:
https://en.wikipedia.org/wiki/Extended_real_number_line#Order_and_topological_properties

You need to be clear what this PR is doing: are you just reorganising code or are you changing the definition of `<=`?

oscarbenjamin replied:
On master we have:
```julia
In [1]: x = Symbol('x', extended_positive=True, infinite=True)                                                                                                

In [2]: y = Symbol('y', extended_positive=True, infinite=True)                                                                                                

In [3]: x <= y                                                                                                                                                
Out[3]: True

In [4]: Le(x, y)                                                                                                                                              
Out[4]: True
```
This PR leaves those two unevaluated.

warnerjon12 replied:
Yeah, I decided to change it back. You're right, it is sometimes helpful to define it that way

warnerjon12 replied:
Alright, it should work like that now

warnerjon12 committed:
Updated test_rel_Infinity in test_relational

warnerjon12 committed:
Updated comparisons in core.numbers to work with infinities correctly

czgdp1807 commented:
@warnerjon12 Can you please add more tests to increase the coverage? Take a look at [this report](https://codecov.io/gh/sympy/sympy/pull/17825/diff).

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -329,98 +329,66 @@ def __complex__(self):
+            diff = self - other
+            if diff is not S.NaN:
+                if op == "<":
+                    test = sympify(diff.is_extended_negative)
```
We should probably use `as_Boolean` rather than `sympify` here.

warnerjon12 replied:
But it could be None

oscarbenjamin replied:
Actually sympify is called below if test is not None so we don't need to call sympify here and can use `as_Boolean` below.

warnerjon12 replied:
`as_Boolean` causes import troubles, but I should just be able to do `S.true if test == True else S.false`

oscarbenjamin commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -390,10 +390,11 @@ def test_issue_10198():
     # is not real the inequality is invalid
-    # assert solve(sin(x) < 2) == (x <= oo)
+    assert solve(sin(x) < 2) == (x <= oo)
+    assert solveset(sin(x) < 2, domain=S.Reals) == S.Reals
```
What has changed here?

warnerjon12 replied:
The original test had that line commented so that it passed with the original behavior being that `x <= oo` would be changed to true. That is not correct unless we know x is real. This tests that the solution is specified by `x <= oo` without it being default true. The second assertion is also part of issue #10047, so I added it also

warnerjon12 replied:
I ended up changing it back to the original test (although I left the second test that I added there), so that it passes with the current implementation of inequality solvers. Only because that issue is beyond the scope of this PR. See issue #17840

warnerjon12 committed:
Fixed some issues related to infinite comparisons

warnerjon12 commented:
The changes are causing some tests to fail, which can mean a few things. My hope is that it will shed light on some other issues that I can work to fix. In any case, I will wait to push again until I get all of the relational and inequality tests passing locally

warnerjon12 committed:
Restored correct behavior of infinite comparisons and fixed some related issues

Comparisons of infinite values will no longer raise ValueError.
Infinities are considered comparable and have their own evalf.
oo.evalf(2) gives Float('inf'), similar for -oo.
If x.is_extended_real, then -oo <= x is S.true and x <= oo is S.true.

warnerjon12 committed:
Fixed an issue with infinite comparisons

warnerjon12 committed:
Reverted test_issue_10047 back to original test

solve(sin(x) < 2) should only unconditionally give true if
it is known that x is extended real. However, this is
implicitly assumed by the inequality solving code, so
I will leave it for now.

warnerjon12 committed:
Fixed an issue when comparing infinite symbols

warnerjon12 committed:
Added tests and fixes for relations on infinite symbols

oscarbenjamin commented on sympy/core/tests/test_relational.py:
```diff
@@ -525,6 +569,9 @@ def test_inequalities_symbol_name_same():
         for b in A:
+            # do not check undefined comparisons of infinite values
+            if a in (-oo, oo) and a == b:
+                continue
```
Is this still needed?

warnerjon12 replied:
Nope. Test still passes when I remove that

warnerjon12 replied:
But it's not worth pushing and resetting the tests for just that

warnerjon12 replied:
I will push that with a couple of other small improvements

warnerjon12 commented:
Well, the tests pass. Let me know if you think anything else is needed. As far as code coverage, it seems like there are a few other changes showing than the ones I made, so it is hard to tell how much I did versus others

sylee957 commented:
I think that a benchmark run should be done. Or if #17828 can be merged, we can check benchmarks on travis after reopening this PR.

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -329,98 +329,80 @@ def __complex__(self):
+                    if diff.is_zero:
+                        test = S.true
+                    elif op == "<=":
+                        test = sympify(diff.is_extended_nonpositive)
```
We don't need to test is_zero separately because if is_zero is True then extended_nonpositive will be True as well.

oscarbenjamin replied:
So this if elif can be flattened

warnerjon12 replied:
Okay, I will commit this and push. I also made the small test change mentioned above

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -329,98 +329,80 @@ def __complex__(self):
+                    return test
+            # oo >= oo, etc.
+            elif self == other:
+                return S.false if op in ("<", ">") else S.true
```
What is this condition needed for? If `self == other` will we not have that `diff` is zero?

warnerjon12 replied:
No, because `oo-oo` is NaN

oscarbenjamin replied:
Isn't that handled above?

warnerjon12 replied:
It seems you are correct. I will push that after I check to see if any other small changes should be made

warnerjon12 replied:
Okay, it should be pushed

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -329,98 +329,80 @@ def __complex__(self):
+            # use float comparison for infinity.
+            # otherwise get stuck in infinite recursion
+            if n2 in (S.Infinity, S.NegativeInfinity):
+                n2 = float(n2)
```
This condition wasn't here before. Why is it needed now?

warnerjon12 replied:
Before, infinity defined its own comparison operators that didn't call _n2. Much of that code duplicated what is done here, so I merged them together. Basically, the problem is that n2 is expected to return either a Float or None for comparable objects. Since infinity is comparable, it doesn't return None, but `Float('inf')` is the same as Infinity which is not actually a Float. So `n2 op 0` calls `_cmp` recursively forever

warnerjon12 replied:
This wouldn't have been needed in older versions where `Float('inf')` was distinct from `S.Infinity`, but I assume that was changed for a reason

warnerjon12 committed:
Small changes to Expr._cmp for efficiency and readability

Also made a small change to a related test.

oscarbenjamin commented on sympy/core/numbers.py:
```diff
@@ -2834,6 +2836,9 @@ def _eval_subs(self, old, new):
             return new
 
+    def evalf(self, prec=None, **options):
+        return Float('inf')
```
Is this just self? Should this not be `_eval_evalf` rather than `evalf`?

warnerjon12 replied:
This was so that when `_n2` calls `evalf` on `Infinity`, it would work right. This is more direct, although I suppose that would also work.

oscarbenjamin replied:
Please understand that I'm just asking questions and no suggesting that the code needs to be done in any particular way.

As I understand it Infinity and `Float('inf')` are the same object. Without this method does something strange happen with evalf?

warnerjon12 replied:
Changing it to `_eval_evalf` causes import troubles. The other alternative would be to check for infinite comparisons before attempting to call `_n2` in `_cmp`, but defining `evalf` for the infinities works well enough. However, I will reimplement it as a wrapper around `_eval_evalf` which I will also redefine. This ensures consistent behavior if `_eval_evalf` is called instead

oscarbenjamin replied:
I don't understand why the method is needed though. Is it not just returning self?

warnerjon12 replied:
Since a few infinite comparisons happen before sympy is fully imported, the call to `evalf` in `_n2` will cause import troubles looking for the definition elsewhere. This is more or less a workaround for that issue

oscarbenjamin replied:
Okay that makes sense

warnerjon12 replied:
I have implemented the wrapper for the infinities, it is pushed

warnerjon12 committed:
Removed a redundant check in Expr._cmp and added _eval_evalf for the infinities

oscarbenjamin commented:
This looks good to me.

The tests haven't passed yet. If/when they do we should run benchmarks against master. Assuming nothing comes up I think this is good to merge.

warnerjon12 commented:
It probably is. I only ran the relational and inequality tests locally, since that is what I was directly dealing with, but I think it will pass

smichr commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -388,12 +388,9 @@ def test_issue_10198():
-    # assert solve(sin(x) < 2) == (x <= oo)
-
     # with PR 16956, (x <= oo) autoevaluates when x is extended_real
     assert solve(sin(x) < 2) == True
```
This really shouldn't be evaluating as True for reasons given in the comment that you deleted. `x` was not declared in this file to be `extended_real` so this shouldn't have evaluated. If it does (and did at the time of the comment) it should be opened as an issue. Please leave the comment and put a "# XXX need to fix" instead of deleting the comment.

oscarbenjamin replied:
This is where it was changed:
https://github.com/sympy/sympy/pull/16956/files#r291032047

warnerjon12 replied:
As I said above, this is beyond the scope of this PR:

> I ended up changing it back to the original test (although I left the second test that I added > there), so that it passes with the current implementation of inequality solvers. Only because > that issue is beyond the scope of this PR. See issue #17840

However, I could restore the test and mark it as expected to fail, perhaps that would be better

oscarbenjamin replied:
I think if you restore that comment and perhaps add to it if there is anything else useful to add then this can be merged

warnerjon12 replied:
Yeah, okay. Works for me

warnerjon12 committed:
Added back a test comment in test_inequalities

warnerjon12 commented:
Should be good for merge, I think

oscarbenjamin commented:
Yeah, looks good. Thanks for this!
